update readme (#10)
DevExpress-Examples/devextreme-project-template-theme-selector: README.md

<!-- default badges list -->
[![](https://img.shields.io/badge/Open_in_DevExpress_Support_Center-FF7200?style=flat-square&logo=DevExpress&logoColor=white)](https://supportcenter.devexpress.com/ticket/details/T1109715)
[![](https://img.shields.io/badge/📖_How_to_use_DevExpress_Examples-e9f6fc?style=flat-square)](https://docs.devexpress.com/GeneralInformation/403183)
[![](https://img.shields.io/badge/💬_Leave_Feedback-feecdd?style=flat-square)](#does-this-example-address-your-development-requirementsobjectives)
<!-- default badges end -->

# Theme Switcher for DevExtreme - How to implement Theme Selector for the DevExtreme Project Template

> This example illustrates how to switch between themes in our Project Template at runtime. We added the [SelectBox](https://js.devexpress.com/Documentation/ApiReference/UI_Components/dxSelectBox/) with the theme list to switch between themes.

Reviewer: ```suggestion
This example demonstrates at-runtime theme switching in our Project Template. You can use a DevExtreme [SelectBox](https://js.devexpress.com/Documentation/ApiReference/UI_Components/dxSelectBox/) within the application to switch between available themes.
```

<!-- default badges list -->
[![](https://img.shields.io/badge/Open_in_DevExpress_Support_Center-FF7200?style=flat-square&logo=DevExpress&logoColor=white)](https://supportcenter.devexpress.com/ticket/details/T1109715)
[![](https://img.shields.io/badge/📖_How_to_use_DevExpress_Examples-e9f6fc?style=flat-square)](https://docs.devexpress.com/GeneralInformation/403183)
[![](https://img.shields.io/badge/💬_Leave_Feedback-feecdd?style=flat-square)](#does-this-example-address-your-development-requirementsobjectives)
<!-- default badges end -->

> # Theme Switcher for DevExtreme - How to Implement a Theme Switcher in the DevExtreme Application Template

Reviewer 2: We've recently checked a similar example for Blazor, so I'll reuse some wordings.

```suggestion
# Theme Switcher for DevExtreme - Implement a Theme Switcher within the DevExtreme Application Template
```

This example demonstrates runtime theme switching in our Application Templates. You can use a DevExtreme [SelectBox](https://js.devexpress.com/Documentation/ApiReference/UI_Components/dxSelectBox/) within the application to switch between available themes.

> DevExtreme Application Templates include two default themes: `base` and `swatch`. These templates also implement SCSS variables to change theme colors.

Reviewer 2: ```suggestion
DevExtreme Application Templates include two default themes: `base` and `swatch`. This example implements a DevExtreme [SelectBox](https://js.devexpress.com/Documentation/ApiReference/UI_Components/dxSelectBox/) designed to switch between themes at runtime.

Our Application Templates contain SCSS variables that allow you to change theme colors. You can also use our predefined [CSS classes](https://js.devexpress.com/Documentation/ApiReference/UI_Components/CSS_Classes/) ([dx-theme-background-color](https://js.devexpress.com/Documentation/ApiReference/UI_Components/CSS_Classes/#dx-theme-background-color), [dx-theme-text-color](https://js.devexpress.com/Documentation/ApiReference/UI_Components/CSS_Classes/#dx-theme-text-color), etc.) to style native HTML elements to maintain a consistent look across your application.
```

To switch between themes, this example implements the following steps:
1) Use the StyleSheet API to disable all loaded `base` stylesheets (excluding the active stylesheet).
2) Use the StyleSheet API to disable all loaded `swatch` stylesheet (excluding the active stylesheet).
3) Update SCSS variables.

> Refer to the following files for the theme switcher logic implementation:

Reviewer 2: Does it make sense to shorten the list as follows?

```suggestion
To switch between themes, use StyleSheet APIs to disable all loaded `base`/`swatch` stylesheets (excluding the active stylesheet) and update SCSS variables. Refer to the following files to review framework-specific implementations:
```

Reviewer: Yeah that is much more straightforward, thank you!

- [Angular Theme Service](Angular/src/app/shared/services/theme.service.ts)
- [Vue Theme Service](Vue/src/services/theme-service.ts)
- [React Theme Service](React/src/contexts/theme.tsx)

This example also uses our predefined [CSS classes](https://js.devexpress.com/Documentation/ApiReference/UI_Components/CSS_Classes/) to customize elements outside of DevExtreme components:

- [dx-theme-background-color](https://js.devexpress.com/Documentation/ApiReference/UI_Components/CSS_Classes/#dx-theme-background-color)

> - [dx-theme-text-color](https://js.devexpress.com/Documentation/ApiReference/UI_Components/CSS_Classes/#dx-theme-text-color)

Reviewer 2: I included this information in the suggestion above, no need to duplicate.

> ### Add a Theme to the Theme Switcher

Reviewer 2: ```suggestion
## Add Themes to the Theme Switcher
```

> 1) Add theme metadata JSON files to `src/themes`. For additional information on theme metadata files, refer to the following topic: [Export Theme Metadata](https://js.devexpress.com/React/Documentation/Guide/Common/DevExtreme_CLI/#ThemeBuilder/Export_Theme_Metadata).

Reviewer 2: ```suggestion
1) Add theme metadata JSON files to `src/themes`. For additional information on theme metadata files, refer to the following help topic: [Export Theme Metadata](https://js.devexpress.com/React/Documentation/Guide/Common/DevExtreme_CLI/#ThemeBuilder/Export_Theme_Metadata).
```

> 2) Add the new metadata file to the command list in **devextreme.json**. The **devextreme.json** file includes a list of themes to build.

Reviewer 2: ```suggestion
2) Add a new metadata file to **devextreme.json** that contains a list of themes to build.
```

Reviewer: To add further info and shorten/simplify this point, I propose the following:
```suggestion
2) Add new metadata files to the command list in **devextreme.json** (as `build-theme` commands).
```

> 3) Run the following command: `npm run build-themes`. This command uses our [ThemeBuilder CLI](https://js.devexpress.com/Documentation/Guide/Common/DevExtreme_CLI/#ThemeBuilder) to generate theme files. You can find built themes in the `outputFile` directory specified in **devextreme.json**.

Reviewer 2: ```suggestion
3) Execute the following command: `npm run build-themes`. This command uses our [ThemeBuilder CLI](https://js.devexpress.com/Documentation/Guide/Common/DevExtreme_CLI/#ThemeBuilder) to generate theme files. You can find these files in the `outputFile` directory (specified in **devextreme.json**).
```

> 4) Add a link to the new theme in **index.html**.

Reviewer 2: ```suggestion
4) Add a link to a new theme in **index.html**.
```

Reviewer: For consistency with point 2:
```suggestion
4) Add links to new themes in **index.html**.
```

> 5) Add a new entry to the array in the `getThemeData` function within the theme service. This array is the data source for the theme switcher SelectBox. This step is required only for base themes.

Reviewer 2: ```suggestion
5) _For base themes only._ Add a new entry to the array in the `getThemeData` function within the theme service. The theme switcher SelectBox uses this array as a data source. 
```

Reviewer: For consistency with point 2; moreover, after reviewing the code once more, getThemeData() returns an array in all examples, so we can simplify as follows:
```suggestion
5) _For base themes only._ Add new entries to the `getThemeData` array within the theme service. The theme switcher SelectBox uses this array as a data source.
```

> This example implements two swatch themes: light and dark. To switch between three or more swatch themes, extend the `applySwatchTheme` method in the theme service.

Reviewer 2: ```suggestion
This example implements two swatch themes (light and dark). To switch between three or more swatch themes, extend the `applySwatchTheme` method in the theme service.
```

> You can also use our ThemeBuilder CLI directly to generate theme files. The following commands generate light and dark swatch themes:

Reviewer 2: ```suggestion
You can also generate theme files using our ThemeBuilder CLI. The following commands generate light and dark swatch themes:
```

[![](https://img.shields.io/badge/💬_Leave_Feedback-feecdd?style=flat-square)](#does-this-example-address-your-development-requirementsobjectives)
<!-- default badges end -->

# Theme Switcher for DevExtreme - How to Implement a Theme Switcher in the DevExtreme Application Template

This example demonstrates runtime theme switching in our Application Templates. You can use a DevExtreme [SelectBox](https://js.devexpress.com/Documentation/ApiReference/UI_Components/dxSelectBox/) within the application to switch between available themes.

DevExtreme Application Templates include two default themes: `base` and `swatch`. These templates also implement SCSS variables to change theme colors.

To switch between themes, this example implements the following steps:
1) Use the StyleSheet API to disable all loaded `base` stylesheets (excluding the active stylesheet).

> 2) Use the StyleSheet API to disable all loaded `swatch` stylesheet.

Author: disable swatch stylesheets and enable the active swatch stylesheet. Similar to base themes

3) Update SCSS variables.

> Review the following files for the theme switcher implementation:

Author: For me, the 'theme switcher' phrase is more about the SelectBox. Mb change to 'theme switching'?
Also, not sure that 'review' is appropriate. May be 'refer'?

- [Angular Theme Service](Angular/src/app/shared/services/theme.service.ts)
- [Vue Theme Service](Vue/src/services/theme-service.ts)
- [React Theme Service](React/src/contexts/theme.tsx)

> Elements such as font and background colors do not change automatically based on the active theme. To customize these elements, you can use predefined [CSS classes](https://js.devexpress.com/Documentation/ApiReference/UI_Components/CSS_Classes/):

Author: Font and background colors are not elements but css attributes. The idea of this sentence is to say that there are elements in views/pages that are not styled by our themes (don't apply theme colors) and users can add the following class to make these elements theme-dependent.

- [dx-theme-background-color](https://js.devexpress.com/Documentation/ApiReference/UI_Components/CSS_Classes/#dx-theme-background-color)
- [dx-theme-text-color](https://js.devexpress.com/Documentation/ApiReference/UI_Components/CSS_Classes/#dx-theme-text-color)

### Add a Theme to the Theme Switcher

> 1) Add theme metadata JSON files to `src/themes`. For additional information on theme metadata exports, refer to the following topic: [Export Theme Metadata](https://js.devexpress.com/React/Documentation/Guide/Common/DevExtreme_CLI/#ThemeBuilder/Export_Theme_Metadata).

Author: 'theme metadata exports'. Even though that topic is about exports, I shared it only because this help topic explains the JSON format/structure of these files. We don't say about exports here but rather about what this file means
